homeassistant/components/media_player/cast.py

```diff
@@ -33,90 +38,164 @@
+    if discovery_info:
+        want_host = (discovery_info.get('host'), discovery_info.get('port'))
+    elif CONF_HOST in config:
+        want_host = (config.get(CONF_HOST), DEFAULT_PORT)
+
+    enable_discovery = False
+    if want_host is None:
+        # We were explicitly told to enable pychromecast discovery.
+        enable_discovery = True
+    elif want_host[1] != DEFAULT_PORT:
+        # We're trying to add a group, so we have to use pychromecast's
+        # discovery to get the correct friendly name.
+        enable_discovery = True
+
+    if enable_discovery:
+        @callback
+        def async_cast_discovered(chromecast):
+            """Callback for when a new chromecast is discovered."""
+            if want_host is not None and \
+                    (chromecast.host, chromecast.port) != want_host:
+                return  # for groups, only add requested device
+            cast_device = _async_create_cast_device(hass, chromecast)
+
+            if cast_device is not None:
+                async_add_devices([cast_device])
+
+        async_dispatcher_connect(hass, SIGNAL_CAST_DISCOVERED,
+                                 async_cast_discovered)
+        # Re-play the callback for all past chromecasts
+        for chromecast in hass.data[KNOWN_CHROMECASTS_KEY].values():
+            async_cast_discovered(chromecast)
+
+        hass.async_add_job(_setup_internal_discovery, hass)
     else:
-        hosts = [tuple(dev[:2]) for dev in pychromecast.discover_chromecasts()
-                 if tuple(dev[:2]) not in known_hosts]
-
-    casts = []
-
-    # get_chromecasts() returns Chromecast objects with the correct friendly
-    # name for grouped devices
-    all_chromecasts = pychromecast.get_chromecasts()
-
-    for host in hosts:
-        (_, port) = host
-        found = [device for device in all_chromecasts
-                 if (device.host, device.port) == host]
-        if found:
-            try:
-                casts.append(CastDevice(found[0]))
-                known_hosts.append(host)
-            except pychromecast.ChromecastConnectionError:
-                pass
-
-        # do not add groups using pychromecast.Chromecast as it leads to names
-        # collision since pychromecast.Chromecast will get device name instead
-        # of group name
-        elif port == DEFAULT_PORT:
-            try:
-                # add the device anyway, get_chromecasts couldn't find it
-                casts.append(CastDevice(pychromecast.Chromecast(*host)))
-                known_hosts.append(host)
-            except pychromecast.ChromecastConnectionError:
-                pass
-
-    add_devices(casts)
+        # Manually add a "normal" Chromecast, we can do that without discovery.
+        try:
+            chromecast = pychromecast.Chromecast(*want_host)
+        except pychromecast.ChromecastConnectionError:
+            _LOGGER.warning("Can't set up chromecast on %s", want_host[0])
+            raise
+        else:
```

Comment:
Since you `raise` above, you don't need to wrap this into an `else`, right?

Reply:
No, I don't :) Will change it